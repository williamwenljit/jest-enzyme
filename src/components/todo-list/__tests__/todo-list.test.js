import React from 'react';
import { shallow, mount } from 'enzyme';
import TodoList from '../';
import TodoItem from '../todo-item';

describe('Todo List', () => {
    it('should handle default props', () => {
        const {
            list,
            inputValue,
            onInputChange,
            onItemAdd,
            onItemRemove,
        } = TodoList.defaultProps;

        expect(list).toMatchObject([]);
        expect(inputValue).toBe('');
        expect(onInputChange).toBeDefined();
        expect(onInputChange).toBeInstanceOf(Function);
        expect(onItemAdd).toBeDefined();
        expect(onItemAdd).toBeInstanceOf(Function);
        expect(onItemRemove).toBeDefined();
        expect(onItemRemove).toBeInstanceOf(Function);
    });

    it('should renders correctly', () => {
        const wrapper = shallow(
            <TodoList
                title="todo list 1"
                list={[{ text: 'todo item 1' }]}
                inputValue=""
            />
        )

        expect(wrapper).toMatchSnapshot();
    });

    it('should contain title', () => {
        const title = 'todo list 1';
        const wrapper = shallow(<TodoList title={title} />);

        expect(wrapper.contains(
            <div className="todo-list__title">
                {title}
            </div>
        )).toEqual(true);
    });

    it('should contain input and button in .todo-list__new', () => {
        const btnText = '新增';
        const wrapper = shallow(<TodoList />);

        expect(wrapper.find('.todo-list__new').containsAllMatchingElements([
            <input />,
            <button>
                {btnText}
            </button>
        ])).toEqual(true);
    });

    it('should handle title props', () => {
        const title = 'todo list 1';
        const wrapper = shallow(<TodoList title={title} />);

        expect(wrapper.find('.todo-list__title').text()).toEqual(title);
    });

    it('should add list item correctly', () => {
        const list = [{ text: 'todo item 1' }];
        const wrapper = shallow(<TodoList />);

        expect(wrapper.find(TodoItem)).toHaveLength(0);
        wrapper.setProps({ list });
        expect(wrapper.find('TodoItem')).toHaveLength(1);
    });

    it('should handle onInputChange', () => {
        const onInputChange = jest.fn();
        const value = 'todo item 1';
        const event = { target: { value } };
        const wrapper = shallow(<TodoList onInputChange={onInputChange} />);

        wrapper.find('.todo-list__new-input').simulate('change', event);
        expect(onInputChange).toHaveBeenCalled();
        expect(onInputChange).toHaveBeenCalledWith(event);
    });

    it('should handle onItemAdd', () => {
        const onItemAdd = jest.fn();
        const text = 'todo item 1';
        const item = { text };
        const wrapper = shallow(<TodoList inputValue={text} onItemAdd={onItemAdd} />);

        wrapper.find('.todo-list__new-button').simulate('click', item);
        expect(onItemAdd).toHaveBeenCalled();
        expect(onItemAdd).toHaveBeenCalledWith(item);
    });

    it('should handle onItemRemove', () => {
        const list = [{ text: 'todo item 1' }, { text: 'todo item 2' }];
        const onItemRemove = jest.fn();
        const index = 1;
        const wrapper = mount(<TodoList list={list} onItemRemove={onItemRemove} />);

        wrapper.find('.todo-item').at(index).find('.todo-item__button').simulate('click', index);
        expect(onItemRemove).toHaveBeenCalled();
        expect(onItemRemove).toHaveBeenCalledWith(index);
    });
});
