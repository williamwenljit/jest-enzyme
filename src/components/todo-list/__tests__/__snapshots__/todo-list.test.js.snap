// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo List should renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    inputValue=""
    list={
      Array [
        Object {
          "text": "todo item 1",
        },
      ]
    }
    onInputChange={[Function]}
    onItemAdd={[Function]}
    onItemRemove={[Function]}
    title="todo list 1"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="todo-list__title"
        >
          todo list 1
        </div>,
        <div
          className="todo-list__new"
        >
          <input
            className="todo-list__new-input"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            className="todo-list__new-button"
            onClick={[Function]}
            type="button"
          >
            新增
          </button>
        </div>,
        <ul
          className="todo-list__content"
        >
          <TodoItem
            index={0}
            onItemRemove={[Function]}
            text="todo item 1"
          />
        </ul>,
      ],
      "className": "todo-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "todo list 1",
          "className": "todo-list__title",
        },
        "ref": null,
        "rendered": "todo list 1",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="todo-list__new-input"
              onChange={[Function]}
              type="text"
              value=""
            />,
            <button
              className="todo-list__new-button"
              onClick={[Function]}
              type="button"
            >
              新增
            </button>,
          ],
          "className": "todo-list__new",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "todo-list__new-input",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "新增",
              "className": "todo-list__new-button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "新增",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TodoItem
              index={0}
              onItemRemove={[Function]}
              text="todo item 1"
            />,
          ],
          "className": "todo-list__content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "function",
            "props": Object {
              "index": 0,
              "onItemRemove": [Function],
              "text": "todo item 1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="todo-list__title"
          >
            todo list 1
          </div>,
          <div
            className="todo-list__new"
          >
            <input
              className="todo-list__new-input"
              onChange={[Function]}
              type="text"
              value=""
            />
            <button
              className="todo-list__new-button"
              onClick={[Function]}
              type="button"
            >
              新增
            </button>
          </div>,
          <ul
            className="todo-list__content"
          >
            <TodoItem
              index={0}
              onItemRemove={[Function]}
              text="todo item 1"
            />
          </ul>,
        ],
        "className": "todo-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "todo list 1",
            "className": "todo-list__title",
          },
          "ref": null,
          "rendered": "todo list 1",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="todo-list__new-input"
                onChange={[Function]}
                type="text"
                value=""
              />,
              <button
                className="todo-list__new-button"
                onClick={[Function]}
                type="button"
              >
                新增
              </button>,
            ],
            "className": "todo-list__new",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "todo-list__new-input",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "新增",
                "className": "todo-list__new-button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": "新增",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TodoItem
                index={0}
                onItemRemove={[Function]}
                text="todo item 1"
              />,
            ],
            "className": "todo-list__content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "index": 0,
                "onItemRemove": [Function],
                "text": "todo item 1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
